File: euler/benchmarks/problem12/csharp/main.cs
using System;
using System.Collections.Generic;

namespace factorbench
{
    class PrimeGenerator
    {
        long n = 3;
        long last = 2;
        Dictionary<long, long> sieve = new Dictionary<long, long>();

        public long Next()
        {
            while (sieve.TryGetValue(n, out var prime))
            {
                sieve.Remove(n);

                var composite = n + prime + prime;
                while (sieve.ContainsKey(composite))
                    composite += prime + prime;
                sieve.Add(composite, prime);

                n = n + 2;
            }

            sieve.Add(n * n, n);
            var r = last;
            last = n;
            n += 2;
            return r;
        }
    }

    class FactorFinder
    {
        Dictionary<long, HashSet<long>> known = new Dictionary<long, HashSet<long>>();
        PrimeGenerator nextPrimes = new PrimeGenerator();
        List<long> knownPrimes = new List<long>();

        public FactorFinder()
        {
            known.Add(1, new HashSet<long> { 1 });
        }

        public List<long> GetPrimeFactors(long of)
        {
            var factors = new List<long>();
            var quotient = of;

            foreach (var prime in knownPrimes)
            {
                if (prime > quotient)
                    return factors;

                var remainder = quotient % prime;
                while (remainder == 0)
                {
                    quotient /= prime;
                    remainder = quotient % prime;
                    factors.Add(prime);
                }
            }

            for (;;)
            {
                var prime = nextPrimes.Next();
                knownPrimes.Add(prime);

                if (prime > quotient)
                    return factors;

                var remainder = quotient % prime;
                while (remainder == 0)
                {
                    quotient /= prime;
                    remainder = quotient % prime;
                    factors.Add(prime);
                }
            }
        }

        public HashSet<long> GetFactors(long of)
        {
            if (known.TryGetValue(of, out var factors))
                return factors;

            factors = new HashSet<long> { 1, of };

            foreach (var prime in GetPrimeFactors(of))
            {
                var factor = of / prime;
                foreach (var subfactor in GetFactors(factor))
                {
                    factors.Add(subfactor);
                }
            }

            known.Add(of, factors);
            return factors;
        }
    }

    static class MainProgram
    {
        static long Solve()
        {
            var finder = new FactorFinder();
            long adder = 0;
            long tn = 0;

            for (;;)
            {
                adder += 1;
                tn += adder;
                var factors = finder.GetFactors(tn);
                if (factors.Count > 1000)
                    return tn;
            }
        }

        public static void Main(string[] args)
        {
            var answer = Solve();
            Console.WriteLine($"Answer: {answer}");
        }
    }
}

File: euler/benchmarks/problem12/swift/main.swift

class PrimeGenerator {
    var n = 3
    var last = 2
    var sieve = [Int: Int]()

    func next() -> Int {
        while let prime = sieve[n] {
            sieve[n] = nil

            var composite: Int = n + prime + prime
            while sieve[composite] != nil {
                composite += prime + prime
            }

            sieve[composite] = prime
            n += 2
        }

        sieve[n*n] = n
        let r = last
        last = n
        n += 2
        return r
    }
}

class FactorFinder {
    var known: [Int: Set<Int>] = [1: [1]]
    var nextPrimes = PrimeGenerator()
    var knownPrimes: [Int] = [Int]()

    func getPrimeFactors(of: Int) -> [Int] {
        var factors = [Int]()
        var quotient = of

        for prime in knownPrimes {
            if prime > quotient {
                return factors
            }

            var remainder = quotient % prime
            while remainder == 0 {
                quotient /= prime
                remainder = quotient % prime
                factors.append(prime)
            }
        }

        while true {
            let prime = nextPrimes.next()
            knownPrimes.append(prime)

            if prime > quotient {
                return factors
            }

            var remainder = quotient % prime
            while (remainder == 0) {
                quotient /= prime
                remainder = quotient % prime
                factors.append(prime)
            }
        }
    }

    func getFactors(of: Int) -> Set<Int> {

        if let existing = known[of] {
            return existing
        }

        let primeFactors = getPrimeFactors(of: of)
        var factors: Set = [1, of]
        for prime in primeFactors {
            let factor = of / prime
            for subfactor in getFactors(of: factor) {
                factors.insert(subfactor)
            }
        }

        known[of] = factors
        return factors
    }
}

func solve() -> Int {
    let finder = FactorFinder()
    var adder = 0
    var tn = 0
    var length = 0

    while length < 1000 {
        adder += 1
        tn += adder

        let factors = finder.getFactors(of: tn)
        length = factors.count
    }

    return tn
}

print("Running")
let answer = solve()
print("Answer: \(answer)")

